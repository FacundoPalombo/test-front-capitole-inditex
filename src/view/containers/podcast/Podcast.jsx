import React from 'react';
import { getPodcastEpisodes } from '../../../services/podcast';
import { Outlet, useLoaderData, useParams, Link } from 'react-router-dom';
import PodcastDetail from './components/PodcastDetail';
import styles from './styles.module.scss';
import { getPodcastChannels } from '../../../services/podcast';

const Podcasts = () => {
  const { podcasts, channels } = useLoaderData();
  const podcast = podcasts.results.find((p) => p.kind === 'podcast');
  const params = useParams();

  /* Quite a bit overkill all this implementation because
    description is on a different api call. But there it is.*/
  const description = channels.feed.entry.find(
    (channel) => channel.id.attributes['im:id'] === params.podcastId
  ).summary.label;

  return (
    <section id="podcasts" className={styles.podcasts}>
      <div data-testid="podcast-detail" className={styles.podcasts__page}>
        <section className={styles.podcasts__detail}>
          <Link
            to={`/podcast/${params.podcastId}`}
            className={styles.podcasts__detail_navigable}
          >
            <PodcastDetail
              image={podcast.artworkUrl600}
              title={podcast.trackName}
              artist={podcast.artistName}
              description={description}
            />
          </Link>
        </section>
        <Outlet />
      </div>
    </section>
  );
};

export async function loader({ params }) {
  const [podcasts, channels] = await Promise.all([
    getPodcastEpisodes({ podcastId: params.podcastId }),
    getPodcastChannels(),
  ]);

  return { podcasts, channels };
}

export default Podcasts;
