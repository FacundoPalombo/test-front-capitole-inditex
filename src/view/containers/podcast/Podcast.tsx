import React from 'react';
import { Outlet, useParams } from 'react-router-dom';
import PodcastDetail from './components/PodcastDetail/PodcastDetail';
import styles from './styles.module.scss';
import {
  episodes as episodesQuery,
  channels as channelsQuery,
} from '@App/queries/podcasts';
import { useIsFetching, useQuery } from '@tanstack/react-query';
import DetailSkeleton from '@components/DetailSkeleton/DetailSkeleton';

const Podcasts = () => {
  const params = useParams();
  const { data: episodes } = useQuery(episodesQuery(params.podcastId));
  const { data: channels } = useQuery(channelsQuery());
  const isLoading = useIsFetching() > 0;

  const episode = episodes?.results?.find((p) => p.kind === 'podcast');

  const description = channels?.feed.entry.find(
    (channel) => channel.id.attributes['im:id'] === params.podcastId
  )?.summary.label;

  // skeleton for early feedback
  if (isLoading) return <DetailSkeleton />;
  return (
    <section id="podcasts" className={styles.podcasts}>
      <div data-testid="podcast-detail" className={styles.podcasts__page}>
        {episode && (
          <PodcastDetail
            image={episode.artworkUrl600!}
            title={episode.trackName!}
            artist={episode.artistName!}
            description={description!}
          />
        )}
        <Outlet />
      </div>
    </section>
  );
};

export const loader =
  (queryClient) =>
  async ({ params }) => {
    const episodesResolvedQuery = episodesQuery(params.podcastId);
    const channelsResolvedQuery = channelsQuery();
    const episodes = await queryClient.ensureQueryData(episodesResolvedQuery);
    const channels = await queryClient.ensureQueryData(channelsResolvedQuery);
    return { channels, episodes };
  };

export default Podcasts;
