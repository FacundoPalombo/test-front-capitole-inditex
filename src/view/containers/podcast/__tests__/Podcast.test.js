import React from 'react';
import Podcast from '../Podcast';
import { render } from '@testing-library/react';

describe('Podcast unit test', () => {
  it('should render ok with props', () => {
    const { asFragment } = render(<Podcast />);
    expect(asFragment()).toMatchSnapshot();
  });
});
