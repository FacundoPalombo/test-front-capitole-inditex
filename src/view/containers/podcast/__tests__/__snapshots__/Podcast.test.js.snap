// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Podcast unit test should render ok with props 1`] = `
<DocumentFragment>
  <section
    id="podcast"
  >
    Podcast
  </section>
</DocumentFragment>
`;
